extensions/vertx-web-common/runtime/src/main/java/io/quarkus/vertx/web/common/runtime/VertxWebCommonRecorder.java

```diff
@@ -0,0 +1,462 @@
+            HttpConfiguration config = new HttpConfiguration();
+            ConfigInstantiator.handleObject(config);
+
+            router = Router.router(VertxCommonRecorder.getVertx());
+            if (hotReplacementHandler != null) {
+                router.route().blockingHandler(hotReplacementHandler);
+            }
+
+            //we can't really do
+            doServerStart(VertxCommonRecorder.getVertx(), config, LaunchMode.DEVELOPMENT);
+
+        } catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public RuntimeValue<Router> initializeRouter(RuntimeValue<Vertx> vertxRuntimeValue, BeanContainer container,
+            HttpConfiguration httpConfiguration, LaunchMode launchMode,
+            boolean startVirtual, boolean startSocket) throws IOException {
+
+        Vertx vertx = vertxRuntimeValue.getValue();
+
+        if (router == null) {
+            router = Router.router(vertx);
+            if (hotReplacementHandler != null) {
+                router.route().handler(hotReplacementHandler);
+            }
+        }
+        // Make it also possible to register the route handlers programmatically
+        Event<Object> event = Arc.container().beanManager().getEvent();
+        event.select(Router.class).fire(router);
+
+        if (startVirtual) {
+            initializeVirtual(vertx);
+        }
+        if (startSocket) {
+            // Start the server
+            if (closeTask == null) {
+                doServerStart(vertx, httpConfiguration, launchMode);
+            }
+        }
+
+        return new RuntimeValue<>(router);
+    }
+
+    public void finalizeRouter(BeanContainer container, Handler<HttpServerRequest> defaultRouteHandler,
+            List<Handler<RoutingContext>> filters, LaunchMode launchMode, ShutdownContext shutdown) {
+        // install the default route at the end
+        if (defaultRouteHandler != null) {
+            //TODO: can we skip the router if no other routes?
+            Route defaultRoute = router.route();
+
+            for (Handler<RoutingContext> filter : filters) {
+                if (filter != null) {
+                    defaultRoute.handler(filter);
+                }
+            }
+
+            defaultRoute.handler(new Handler<RoutingContext>() {
+                @Override
+                public void handle(RoutingContext event) {
+                    defaultRouteHandler.handle(event.request());
+                }
+            });
+        }
+
+        if (launchMode == LaunchMode.DEVELOPMENT) {
+            shutdown.addShutdownTask(new Runnable() {
+                @Override
+                public void run() {
+                    List<Route> routes = router.getRoutes();
+                    // if the hot replacement handler has been installed, we keep it
+                    for (int i = (hotReplacementHandler != null ? 1 : 0); i < routes.size(); i++) {
+                        routes.get(i).remove();
+                    }
```

Comment:
I now clean up the routes properly by only keeping the hot replacement one if we have installed it.